### PR TITLE
Output directory is created if it doesn't exists. 

### PR DESCRIPTION
**Pull request by SebastianOsuna**
Pull Request for #574 .

Changes are on 'output-directory' branch.

:)

### CODE REVIEW

**Comment by kevva** [comment]
-1. Just replace `fs` with `fs-extra` and use the `fs.outputFile()` function instead of `fs.writeFile()`.

**Comment by am11** [comment]
I don't think we need extra dependency (which further depends on three more dependencies) for a little chore like this.

**Comment by kevva** [comment]
Well, this needs rewriting then. Code looks horrible.

File: lib/render.js [hunk 1]
```diff
@@ -47,16 +49,38 @@ module.exports = function(options, emitter) {
 
     emitter.emit('warn', chalk.green('Rendering Complete, saving .css file...'));
 
-    fs.writeFile(options.dest, result.css, function(err) {
-      if (err) {
-        return emitter.emit('error', chalk.red(err));
-      }
+    // Check if output directory exists
+    var dest_dir = path.dirname(path.normalize(options.dest));
+    var write_out = function() {
```
Comment (kevva) [inline lib/render.js]: Functions are written using camelCase like `function writeOut()` and not using `var`.

File: lib/render.js [hunk 2]
```diff
@@ -47,16 +49,38 @@ module.exports = function(options, emitter) {
...
-      emitter.emit('warn', chalk.green('Wrote CSS to ' + options.dest));
-      emitter.emit('write', err, options.dest, result.css);
-      done();
+        emitter.emit('warn', chalk.green('Wrote CSS to ' + options.dest));
+        emitter.emit('write', err, options.dest, result.css);
+        done();
+      });
+    };
+
+    fs.exists(dest_dir, function(exists) {
```
Comment (kevva) [inline lib/render.js]: Remove this and use this directly:

```js
mkdirp(path.dirname(options.dest), function(err) {
  if (err) {
    return emitter.emit('error', chalk.red(err));
  }

  fs.writeFile(options.dest, result.css, function(err) {
    if (err) {
      return emitter.emit('error', chalk.red(err));
    }

    emitter.emit('warn', chalk.green('Wrote CSS to ' + options.dest));
    emitter.emit('write', err, options.dest, result.css);
    done();
  });
});
```

File: test/cli.js [hunk 3]
```diff
@@ -225,6 +225,30 @@ describe('cli', function() {
...
 
+  describe('node-sass in.scss --output path/to/file/out.css', function() {
+    it('should create the output directory', function(done) {
+      var src = fixture('output-directory/index.scss');
+      var dest = fixture('output-directory/path/to/file/index.css');
+      var bin = spawn(cli, [src, '--output', path.dirname(dest)]);
+
+      bin.on('close', function() {
+        assert(fs.existsSync(path.dirname(dest)));
+        // Delete file
```
Comment (kevva) [inline test/cli.js]: Remove the comments.

**Comment by am11** [comment]
Agree, this would suffice:

```javascript
mkdirp(path.dirname(options.dest), function(err) { 
  if (err) { 
    return emitter.emit('error', chalk.red(err)); 
  }

  fs.writeFile(options.dest, result.css, function(err) { 
    if (err) { 
      return emitter.emit('error', chalk.red(err)); 
    }

    emitter.emit('warn', chalk.green('Wrote CSS to ' + options.dest)); 
    emitter.emit('write', err, options.dest, result.css); 
    done();
  });
});
```

i.e. no need for separate function and `path.normalize()`.

**Comment by am11** [comment]
@SebastianOsuna, we are readying for vNext, could you make those changes and rebase soon?

**Comment by SebastianOsuna** [comment]
Fixes done. I'm not sure if the rebase was successful, I'm not an expert with git yet, sorry.

**Comment by am11** [comment]
@SebastianOsuna, 

```bash
git pull --rebase
git rebase -i HEAD 5

# the editor will open
# change the word 'pick' to 'f', for each
# uncommented line after the one with text:
# "Output directory is created if it doesn't exists. "
# save and close the editor

# now forcefully rewrite the history:
git push -f
```

That's it! :)

**Comment by SebastianOsuna** [comment]
Rebase done with the commands provided. Sorry for the delay.

**Comment by am11** [comment]
@SebastianOsuna, the outcome should be one commit. But right now you have 8 commits.

So lets try this:

```
git remote add node-sass https://github.com/sass/node-sass
git pull --rebase node-sass master
git push -f
```

This will clean those other commits your PR is showing.

Then we will rebase in interactive mode (later, once you rebase against node-sass master).

**Comment by SebastianOsuna** [comment]
Ok, did that but still 2 commits and both duplicated :(

**Comment by am11** [comment]
That's ok. Now you can run:

```bash
git pull --rebase
git rebase -i HEAD~4
```

And then in editor, except for the first line change `pick` to `f` for the next three lines, save and close the editor. Finally, `git push -f`, to rewrite the history.

**Comment by SebastianOsuna** [comment]
Sorry, but I can't seem to make it work. Each time I try the fixup, the log doesn't change. I even tried with the squash option, but no success.

**Comment by am11** [comment]
If you are on Windows, you would probably get a notepad opened after you do `git rebase -i HEAD~4` with the following:

```
pick 099e501 Output directory is created if it doesn't exists.
pick b180fb5 Output directory is created if it doesn't exists.  
pick c2d56dc Code styling fixes.
pick 91f2012 Code styling fixes.
```

Change it to:

```
pick 099e501 Output directory is created if it doesn't exists.
f b180fb5 Output directory is created if it doesn't exists.  
f c2d56dc Code styling fixes.
f 91f2012 Code styling fixes.
```

Ctrl+S and close notepad and run `git push -f`.

Does this work?

**Comment by SebastianOsuna** [comment]
Was closing vim with the wrong command, sorry.

**Comment by am11** [comment]
Perfect. Thanks for your patience and contribution! :+1:

**Pull request merged**